src/api.cu

```diff
@@ -167,128 +170,163 @@ TRITONBACKEND_ModelInstanceExecute(
 {
   uint64_t all_start_time =
       std::chrono::steady_clock::now().time_since_epoch().count();
-  std::size_t total_inference_count = 0;
-  std::vector<TRITONBACKEND_Response*> responses;
+
+  std::vector<TRITONBACKEND_Request*> requests(
+      raw_requests, raw_requests + request_count);
+
+  auto instance_state = get_instance_state<ModelInstanceState>(*instance);
+  auto model_state = instance_state->StateForModel();
+  auto target_memory = get_native_memory_for_instance(instance_state->Kind());
+
+  // Get all input data and bail out gracefully if this fails
+  std::vector<InputBatch<float>> input_batches;
   try {
-    auto instance_state = get_instance_state<ModelInstanceState>(*instance);
-    ModelState* model_state = instance_state->StateForModel();
-    auto target_memory = get_native_memory_for_instance(instance_state->Kind());
+    input_batches = get_input_batches<float>(
+        static_cast<uint32_t>(0), requests, target_memory,
+        instance_state->get_raft_handle());
+  }
+  catch (TritonException& setup_err) {
+    return setup_err.error();
+  }
 
-    std::vector<TRITONBACKEND_Request*> requests(
-        raw_requests, raw_requests + request_count);
+  // BEGIN: Lambda definitions for processing individual batches
+
+  /* Given an input batch and output batch, perform prediction on input,
+   * storing results in output and returning timing information. Return value
+   * is a std::optional containing a pair of uint64_t representing nanoseconds
+   * since the epoch to the beginning and end of prediction respectively. If
+   * std::optional is std::nullopt, prediction has failed and timing values
+   * would have no meaning */
+  auto instance_predict = [&](decltype(*input_batches.begin())& input_batch,
```

Comment:
The generic answer is in order to follow [Almost Always Auto](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/) style, but the more specific advantage that it gives us here is that at this point in the code, we do not actually care what the type of `input_batch` is, only that it is what is stored in `input_batches` and accepted by `instance_state->predict`. If the interface changes later on, and `input_batches` now contained something that could be converted to an `InputBatch<float>` implicitly but which was not an `InputBatch<float>`, we might accidentally start incurring the cost of an unnecessary conversion. Using `decltype` here signals that our only goal is to "hook up the pipes" without insisting on a particular type.

That does raise the question of course of why I explicitly named the type for `output_batch`, and the answer is I have no idea! I'll fix that up.

Reply:
Oh, I remember why. We don't have an `output_batch` until line 244, and I felt like declaring the lambda earlier made for a cleaner/clearer structure. What do you think? Okay as is, or should I reconsider?

Reply:
Otherwise, in lambas, I believe you can use `auto` directly

Reply:
Declaring the lambda here instead of at the same scope than `output_batch` is definitely cleaner, so I'm okay with this solution

Reply:
> Otherwise, in lambas, I believe you can use auto directly

D'oh! You're quite right. I think I noticed that we needed the `decltype` at lines 233-234 and then got hung up on using it in all the other lambdas.

In terms of being type-agnostic, my argument is not that that lambda *is* type agnostic but it is "just" a restructuring of the existing code. In the original version, that logic would have been in a context where the type is already known, so moving it into a lambda for clearer/cleaner organization should not introduce the need for a new type declaration.

Speaking more generally, I'd argue that `auto` is not about type agnosticism but about deferring type determination to some other source of truth. The following is not really type agnostic, but it's a sensible use of auto:
```c++
std::vector<int> v = {0, 1, 2, 3, 4};
auto i = std::begin(v);
std::cout << *(i++) << "\n";
```
Here, we care very much about the type of `i`. It must be incrementable, and the return type of that incrementation must be dereferencable, so we are depending on the existence of certain members. The benefit that `auto` brings is that we are deferring to `std::begin` (and  from there to `std::vector`) as the "source of truth" about what `i` should be. If that type doesn't have the members we want, then it will just fail to compile, and we know that our source of truth gave us a type that behaves differently than we expected rather than attempting to insist that it is something else using another source of truth. In this specific case, the lambda is ultimately deferring to `get_input_batches` to tell us the type of an input batch, since that function is the source of those objects.

Sorry, that was an excessively-detailed answer, but it was as much for me to think through this myself as anything :smile:. In an effort to make sure that this code is at least fully self-consistent about how I'm using `auto`, I've tweaked a couple more things to get to full AAA-style in `TRITONBACKEND_ModelInstanceExecute`. If we still don't like the implementation after that change, we can look at a more significant restructuring.

Reply:
Love the detailed answer! I don't have much more to add, you're absolutely right. `auto` defers the type checking, but we're still in compiler error land if an input is tried to be passed to the lambda whose type does not work with the contents of the lambda